File: android/dotrino-native/src/main/java/com/dotrino/sdk/ProxyConnection.kt
package com.dotrino.sdk

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

/**
 * ONE connection to the proxy, speaking the same frames as `@dotrino/proxy-client`
 * (`src/client.js`): `connected` gives the token, `identify` binds my key, a directed
 * message goes `{ to_publickey, message }` and arrives as `{ type:'message', from, message }`.
 *
 * It does not reconnect by itself: whoever owns it (one per account, while the approvals
 * screen is open) decides what a dropped connection means. A dead socket fails every
 * pending request with its reason instead of leaving it to time out.
 */
class ProxyConnection(private val url: String) {
    companion object {
        private val http: OkHttpClient by lazy {
            OkHttpClient.Builder().pingInterval(20, TimeUnit.SECONDS).build()
        }
        private val json = Json { ignoreUnknownKeys = true }
    }

    /** A directed message: who sent it (connection token and, if identified, key) and its payload. */
    data class Incoming(val from: String?, val fromPubkey: String?, val payload: JsonObject)

    class ProxyError(message: String, val code: String) : Exception(message)

    private var ws: WebSocket? = null
    private val connected = CompletableDeferred<String>()
    private val pending = ConcurrentHashMap<String, CompletableDeferred<JsonObject>>()
    private val nextId = AtomicInteger(1)
    private val listeners = CopyOnWriteArrayList<(Incoming) -> Unit>()
    @Volatile var closed: String? = null; private set
    var token: String? = null; private set

    /** The audience that goes inside `identify`: the proxy URL without trailing slashes. */
    val audience: String get() = url.trimEnd('/')

    suspend fun connect(timeoutMs: Long = 10_000): String {
        ws = http.newWebSocket(Request.Builder().url(url).build(), object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) = handle(text)
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) = die("transport: ${t.message}")
            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) = die("closed: $code $reason")
        })
        return withTimeout(timeoutMs) { connected.await() }
    }

    fun onMessage(l: (Incoming) -> Unit): () -> Unit { listeners.add(l); return { listeners.remove(l) } }

    fun close() { ws?.close(1000, "bye"); die("closed by us") }

    private fun die(reason: String) {
        if (closed != null) return
        closed = reason
        val e = ProxyError(reason, "disconnected")
        connected.completeExceptionally(e)
        pending.values.forEach { it.completeExceptionally(e) }
        pending.clear()
    }

    private fun handle(text: String) {
        val o = try { json.parseToJsonElement(text) as? JsonObject } catch (_: Exception) { null } ?: return
        val type = o["type"]?.jsonPrimitive?.content
        val id = o["id"]?.jsonPrimitive?.content
        when (type) {
            "connected" -> {
                val t = (o["instance"] ?: o["token"])?.jsonPrimitive?.content
                if (t == null) { die("connected without a token"); return }
                token = t; connected.complete(t)
            }
            "message" -> {
                val raw = o["message"] ?: return
                val payload = when (raw) {
                    is JsonObject -> raw
                    is JsonPrimitive -> try { json.parseToJsonElement(raw.content) as? JsonObject } catch (_: Exception) { null }
                    else -> null
                } ?: return
                val inc = Incoming(o["from"]?.jsonPrimitive?.content, o["from_publickey"]?.jsonPrimitive?.content, payload)
                listeners.forEach { runCatching { it(inc) } }
            }
            "error" -> if (id != null) pending.remove(id)?.completeExceptionally(
                ProxyError(o["error"]?.jsonPrimitive?.content ?: "proxy error", o["code"]?.jsonPrimitive?.content ?: "proxy-error"))
            else -> if (id != null) pending.remove(id)?.complete(o)
        }
    }

    private fun send(frame: JsonObject) {
        closed?.let { throw ProxyError(it, "disconnected") }
        val w = ws ?: throw ProxyError("not connected", "disconnected")
        if (!w.send(frame.toString())) throw ProxyError("could not send: socket closing", "disconnected")
    }

    /** A request with an `id` whose answer comes back with the same `id`. */
    private suspend fun request(frame: JsonObject, timeoutMs: Long = 10_000): JsonObject {
        val id = "req_${nextId.getAndIncrement()}"
        val d = CompletableDeferred<JsonObject>()
        pending[id] = d
        try {
            send(JsonObject(frame + ("id" to JsonPrimitive(id))))
            return withTimeout(timeoutMs) { d.await() }
        } finally { pending.remove(id) }
    }

    /**
     * Binds this connection to my key: messages written to it arrive here live, and what was
     * queued while I was away gets delivered. The token is the challenge of this connection
     * and goes inside what is signed; `aud` says who it is meant for.
     */
    suspend fun identify(keys: DeviceKeys) {
        val t = token ?: throw ProxyError("identify before connecting", "disconnected")
        val data = buildJsonObject {
            put("op", "identify"); put("aud", audience); put("publickey", keys.publickey)
            put("token", t); put("ts", System.currentTimeMillis())
        }
        request(buildJsonObject {
            put("type", "identify"); put("data", data); put("signature", keys.sign(Canonical.stringify(data)))
        })
    }

    /** Registers this phone's FCM token under my key: the proxy rings it when something is queued for me. */
    suspend fun registerPushToken(keys: DeviceKeys, fcmToken: String) {
        val sub = buildJsonObject { put("kind", "fcm"); put("token", fcmToken) }.toString()
        val data = buildJsonObject {
            put("op", "push-subscribe"); put("publickey", keys.publickey); put("subscription", sub)
            put("ts", System.currentTimeMillis())
        }
        request(buildJsonObject {
            put("type", "push-subscribe"); put("data", data); put("signature", keys.sign(Canonical.stringify(data)))
        })
    }

    /** A directed message to a key. The payload travels as a JSON string, like the JS client sends it. */
    fun sendByPubkey(to: String, payload: JsonObject) {
        send(buildJsonObject {
            put("to_publickey", buildJsonArray { add(JsonPrimitive(to)) })
            put("message", payload.toString())
        })
    }
}

File: ios/DotrinoNative/ProxyConnection.swift
import Foundation

public struct ProxyError: Error, CustomStringConvertible {
    public let description: String
    public let code: String
    public init(_ d: String, code: String) { description = d; self.code = code }
}

/// A value that arrives once, awaited with a deadline. The first `finish` wins; later ones
/// (a late answer, a timeout after the answer) are ignored.
final class OneShot<T>: @unchecked Sendable {
    private let lock = NSLock()
    private var cont: CheckedContinuation<T, Error>?
    private var result: Result<T, Error>?

    func finish(_ r: Result<T, Error>) {
        lock.lock()
        if result != nil { lock.unlock(); return }
        result = r
        let c = cont; cont = nil
        lock.unlock()
        c?.resume(with: r)
    }

    func wait(timeout: TimeInterval, onTimeout: @autoclosure @escaping () -> Error) async throws -> T {
        try await withCheckedThrowingContinuation { (c: CheckedContinuation<T, Error>) in
            lock.lock()
            if let r = result { lock.unlock(); c.resume(with: r); return }
            cont = c
            lock.unlock()
            Task { [weak self] in
                try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
                self?.finish(.failure(onTimeout()))
            }
        }
    }
}

/// ONE connection to the proxy, speaking the same frames as `@dotrino/proxy-client`
/// (`src/client.js`): `connected` gives the token, `identify` binds my key, a directed
/// message goes `{ to_publickey, message }` and arrives as `{ type:'message', from, message }`.
///
/// It does not reconnect by itself: whoever owns it (one per account, while the requests
/// screen is open) decides what a dropped connection means. A dead socket fails every
/// pending request with its reason instead of leaving it to time out.
public final class ProxyConnection: NSObject, URLSessionWebSocketDelegate, @unchecked Sendable {
    /// A directed message: who sent it (connection token and, if identified, key) and its payload.
    public struct Incoming: Sendable {
        public let from: String?
        public let fromPubkey: String?
        public let payload: JSON
    }

    private let url: URL
    private let lock = NSLock()
    private var session: URLSession!
    private var ws: URLSessionWebSocketTask?
    private let connected = OneShot<String>()
    private var pending: [String: OneShot<JSON>] = [:]
    private var nextId = 1
    private var listeners: [UUID: (Incoming) -> Void] = [:]
    private var _closed: String?
    public private(set) var token: String?

    public var closed: String? { lock.lock(); defer { lock.unlock() }; return _closed }

    /// The audience that goes inside `identify`: the proxy URL without trailing slashes.
    public let audience: String

    public init(_ urlString: String) throws {
        guard let u = URL(string: urlString) else { throw ProxyError("invalid proxy url", code: "bad-url") }
        url = u
        var a = urlString
        while a.hasSuffix("/") { a.removeLast() }
        audience = a
        super.init()
        session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
    }

    @discardableResult
    public func connect(timeout: TimeInterval = 10) async throws -> String {
        let t = session.webSocketTask(with: url)
        t.maximumMessageSize = 4 * 1024 * 1024
        lock.lock(); ws = t; lock.unlock()
        t.resume()
        receive(t)
        ping()
        return try await connected.wait(timeout: timeout, onTimeout: ProxyError("the proxy did not greet in time", code: "timeout"))
    }

    public func onMessage(_ l: @escaping (Incoming) -> Void) -> () -> Void {
        let key = UUID()
        lock.lock(); listeners[key] = l; lock.unlock()
        return { [weak self] in self?.lock.lock(); self?.listeners[key] = nil; self?.lock.unlock() }
    }

    public func close() {
        ws?.cancel(with: .normalClosure, reason: nil)
        die("closed by us")
    }

    private func die(_ reason: String) {
        lock.lock()
        if _closed != nil { lock.unlock(); return }
        _closed = reason
        let all = pending.values
        pending.removeAll()
        lock.unlock()
        let e = ProxyError(reason, code: "disconnected")
        connected.finish(.failure(e))
        all.forEach { $0.finish(.failure(e)) }
        // The session holds its delegate (this object) until invalidated: without this every
        // dropped connection would stay in memory.
        session.invalidateAndCancel()
    }

    private func receive(_ t: URLSessionWebSocketTask) {
        t.receive { [weak self] r in
            guard let self else { return }
            switch r {
            case .failure(let e): self.die("transport: \(e.localizedDescription)")
            case .success(let m):
                switch m {
                case .string(let s): self.handle(s)
                case .data(let d): if let s = String(data: d, encoding: .utf8) { self.handle(s) }
                @unknown default: break
                }
                if self.closed == nil { self.receive(t) }
            }
        }
    }

    /// Keeps NATs and the proxy from dropping an idle socket (okhttp's `pingInterval` on Android).
    private func ping() {
        Task { [weak self] in
            while true {
                try? await Task.sleep(nanoseconds: 20_000_000_000)
                guard let self, self.closed == nil, let t = self.ws else { return }
                t.sendPing { e in if let e { self.die("ping: \(e.localizedDescription)") } }
            }
        }
    }

    public func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
        die("closed: \(closeCode.rawValue) \(reason.flatMap { String(data: $0, encoding: .utf8) } ?? "")")
    }

    private func handle(_ text: String) {
        guard let o = try? JSON.parse(text), o.object != nil else { return }
        let type = o["type"]?.string
        let id = o["id"]?.string
        switch type {
        case "connected":
            guard let t = (o["instance"] ?? o["token"])?.string else { die("connected without a token"); return }
            lock.lock(); token = t; lock.unlock()
            connected.finish(.success(t))
        case "message":
            let payload: JSON?
            switch o["message"] {
            case .object?: payload = o["message"]
            case .string(let s)?: payload = (try? JSON.parse(s)).flatMap { $0.object != nil ? $0 : nil }
            default: payload = nil
            }
            guard let payload else { return }
            let inc = Incoming(from: o["from"]?.string, fromPubkey: o["from_publickey"]?.string, payload: payload)
            lock.lock(); let ls = Array(listeners.values); lock.unlock()
            ls.forEach { $0(inc) }
        case "error":
            if let id, let p = take(id) {
                p.finish(.failure(ProxyError(o["error"]?.string ?? "proxy error", code: o["code"]?.string ?? "proxy-error")))
            }
        default:
            if let id, let p = take(id) { p.finish(.success(o)) }
        }
    }

    private func take(_ id: String) -> OneShot<JSON>? {
        lock.lock(); defer { lock.unlock() }
        return pending.removeValue(forKey: id)
    }

    private func send(_ frame: JSON) throws {
        if let c = closed { throw ProxyError(c, code: "disconnected") }
        guard let t = ws else { throw ProxyError("not connected", code: "disconnected") }
        t.send(.string(frame.text)) { [weak self] e in if let e { self?.die("send: \(e.localizedDescription)") } }
    }

    /// A request with an `id` whose answer comes back with the same `id`.
    @discardableResult
    private func request(_ frame: [String: JSON], timeout: TimeInterval = 10) async throws -> JSON {
        let p = OneShot<JSON>()
        lock.lock()
        let id = "req_\(nextId)"; nextId += 1
        pending[id] = p
        lock.unlock()
        defer { _ = take(id) }
        var f = frame
        f["id"] = .string(id)
        try send(.object(f))
        return try await p.wait(timeout: timeout, onTimeout: ProxyError("the proxy did not answer \(frame["type"]?.string ?? "")", code: "timeout"))
    }

    /// Binds this connection to my key: messages written to it arrive here live, and what was
    /// queued while I was away gets delivered. The token is the challenge of this connection
    /// and goes inside what is signed; `aud` says who it is meant for.
    public func identify(_ keys: DeviceKeys) async throws {
        guard let t = token else { throw ProxyError("identify before connecting", code: "disconnected") }
        let data: JSON = ["op": "identify", "aud": .string(audience), "publickey": .string(keys.publickey),
                          "token": .string(t), "ts": .int(nowMs())]
        try await request(["type": "identify", "data": data, "signature": .string(keys.sign(Canonical.stringify(data)))])
    }

    /// A directed message to a key. The payload travels as a JSON string, like the JS client sends it.
    public func sendByPubkey(_ to: String, _ payload: JSON) throws {
        try send(["to_publickey": [.string(to)], "message": .string(payload.text)])
    }
}
